src/Sparrow.Server/Utils/DiskStatsGetter.cs

```diff
@@ -0,0 +1,137 @@
+using System.Threading.Tasks;
+using Sparrow.Logging;
+using Sparrow.Server.Platform;
+
+namespace Sparrow.Server.Utils
+{
+    public interface IDiskStatsGetter
+    {
+        DiskStatsResult Get(string drive);
+        Task<DiskStatsResult> GetAsync(string drive);
+    }
+
+    internal class DiskStatsGetter : IDiskStatsGetter
+    {
+        private readonly TimeSpan _minInterval;
+        private static readonly Logger Logger = LoggingSource.Instance.GetLogger("Server", typeof(DiskStatsGetter).FullName);
+
+        private readonly ConcurrentDictionary<string, Task<PrevInfo>> _previousInfo = new ConcurrentDictionary<string, Task<PrevInfo>>();
+
+        public DiskStatsGetter(TimeSpan minInterval)
+        {
+            _minInterval = minInterval;
+        }
+
+        public DiskStatsResult Get(string drive) => GetAsync(drive).ConfigureAwait(false).GetAwaiter().GetResult();
+        
+        public async Task<DiskStatsResult> GetAsync(string drive)
+        {
+            if (drive == null)
+                return null;
+
+            var start = DateTime.UtcNow;
+            State state = null;
+            while (true)
+            {
+                if (_previousInfo.TryGetValue(drive, out var prevTask) == false)
+                {
+                    state = new State {Drive = drive};
+                    var task = new Task<PrevInfo>(GetStats, state);
+                    if (_previousInfo.TryAdd(drive, task) == false)
+                        continue;
+
+                    task.Start();
+                    return null;
+                }
+
+                var prev = await prevTask.ConfigureAwait(false);
+                if (prev == null)
+                    return null;
+                
+                var diff = DateTime.UtcNow - prev.Raw.Time;
+                if (start < prev.Raw.Time || diff < _minInterval)
+                    return prev.Calculated;
+
+                state ??= new State {Drive = drive};
+                state.Prev = prev;
+
+                var calculateTask = new Task<PrevInfo>(CalculateStats, state);
+                if (_previousInfo.TryUpdate(drive, calculateTask, prevTask) == false)
+                    continue;
+
+                calculateTask.Start();
+                return (await calculateTask.ConfigureAwait(false)).Calculated;
+            }
+        }
+
+        private PrevInfo GetStats(object o)
+        {
+            var state = (State)o;
+            var currentInfo = GetDiskInfo(state.Drive);
+            return currentInfo != null ? new PrevInfo {Raw = currentInfo} : null;
+        }
+
+        private PrevInfo CalculateStats(object o)
+        {
+            var state = (State)o;
+            var currentInfo = GetDiskInfo(state.Drive);
+            if (currentInfo == null)
+                return null;
+
+            var diff = (currentInfo.Time - state.Prev.Raw.Time).TotalSeconds;
+            var read = (currentInfo.ReadIOs - state.Prev.Raw.ReadIOs) / diff;
```

Comment:
Not that on for big uptimes these values will overflow and start from 0.
https://serverfault.com/questions/645331/what-is-the-counter-type-for-linuxs-sys-block-device-stat

Reply:
Only on 32 bits.